pkg/segment/aggregations/segaggs.go

```diff
@@ -196,26 +206,247 @@ func performAggOnResult(nodeResult *structs.NodeResult, agg *structs.QueryAggreg
 	return nil
 }
 
-func performMaxRows(nodeResult *structs.NodeResult, aggs *structs.QueryAggregators, maxRows uint64, recs map[string]map[string]interface{}) error {
+// only called when headExpr has BoolExpr
+func performConditionalHeadOnHistogram(nodeResult *structs.NodeResult, headExpr *structs.HeadExpr) error {
+	fieldsInExpr := headExpr.BoolExpr.GetFields()
+	fieldToValue := make(map[string]segutils.CValueEnclosure, 0)
+
+	for _, aggregationResult := range nodeResult.Histogram {
+		newResults := make([]*structs.BucketResult, 0)
+
+		if !headExpr.Done {
+			for rowIndex, bucketResult := range aggregationResult.Results {
+				// Get the values of all the necessary fields.
+				err := getAggregationResultFieldValues(fieldToValue, fieldsInExpr, aggregationResult, rowIndex)
+				if err != nil {
+					return fmt.Errorf("performConditionalHeadOnHistogram: error while getting agg result fields values, err: %v", err)
+				}
+
+				// Evaluate the expression to a value.
+				result, err := headExpr.BoolExpr.Evaluate(fieldToValue)
+				if err != nil {
+					nullFields, err := headExpr.BoolExpr.GetNullFields(fieldToValue)
+					if err == nil && len(nullFields) > 0 {
+						// evaluation failed due to null fields
```

Comment:
Can there be a case where err is `nil` and `len(nullFields)` is also zero?

Then since the second condition is failed, it will go to else block and return with Error.

I believe it is better if you refactor it so that you check `err != null` condition first and return the error if it is not null and then check for `len(nullFields) == 0` and then return with appropriate message saying similar to `Null Fields is Empty`.

Also, generally it is always advised to check the `err != nil` check and then do the rest of the handling.

Reply:
addressed in the latest commit. When err is nil and len(nullFields) == 0, it means that evaluation failed due to some other reason so we would want to return an error that we received before.